trinity/db/pipe.py

```diff
@@ -35,69 +43,136 @@ def run_command(pipe_db, *params, command):
+            logger.info("Breaking connection loop due to error: %s", err)
             break
 
-        try:
-            req_id, method, *params = request
-        except (TypeError, ValueError) as err:
-            # TypeError: request is non iterable
-            # ValueError: request is less than length-2
-            pipe.send(err)
 
-        try:
-            if method == GET:
-                key = params[0]
-                logger.debug('GET: %s', key)
-
-                try:
-                    pipe.send([req_id, db.get(key)])
-                except KeyError as err:
-                    pipe.send([req_id, err])
-            elif method == SET:
-                key, value = params
-                logger.debug('SET: %s -> %s', key, value)
-                try:
-                    pipe.send([req_id, db.set(key, value)])
-                except KeyError as err:
-                    pipe.send([req_id, err])
-            elif method == EXISTS:
-                key = params[0]
-                logger.debug('EXISTS: %s', key)
-                try:
-                    pipe.send([req_id, db.exists(key)])
-                except KeyError as err:
-                    pipe.send([req_id, err])
-            elif method == DELETE:
-                key = params[0]
-                logger.debug('DELETE: %s', key)
-                try:
-                    pipe.send([req_id, db.delete(key)])
-                except KeyError as err:
-                    pipe.send([req_id, err])
-            else:
-                logger.error("Got unknown method: %s: %s", method, params)
-                pipe.send(Exception("Invalid request method: {0}".format(method)))
-        except Exception as err:
-            # Push the failure out to whatever process was trying to interact
-            # with the database.
-            pipe.send(err)
+def db_server(db, ipc_path):
+    logger = logging.getLogger('trinity.db.pipe.db_server')
+
+    logger.info('Starting DB Process')
+
+    # TODO: this check should be moved higher in the stack so that things fail
+    # prior to spinning up the database server.
+    if os.path.exists(ipc_path):
+        raise OSError(
+            "Found existing IPC socket at {0}.  Either there is another running "
+            "process using this socket or the client **may** have failed to "
+            "failed to shut down cleanly."
+        )
+
+    with Listener(ipc_path) as listener:
+        connections = []
+
+        # We spin up a thread which accepts incomming connection requests and
+        # adds them to the list of connection objects.
+        thread = threading.Thread(
+            target=_accept_connections,
+            args=(listener, connections),
+            daemon=True,
+        )
+        thread.start()
+
+        while True:
+            # If we have no connections, sleep for a moment, otherwise `wait`
+            # blocks indefinitely on an empty list.
+            if not connections:
+                time.sleep(0.1)
```

Comment:
Hm, this sleep and check stuff makes me think we might be better off with an `asyncio.start_unix_server(...)` style server.

Like https://github.com/ethereum/py-evm/blob/master/evm/rpc/ipc.py#L99

Doesn't seem urgent if this works, though. Just an idea.

Reply:
I looked into this to some degree.  What I couldn't figure out was how to get the friendly API that `multiprocessing.Connection` has of automatic pickling and unpickling of objects which is key to making this work without a lot more overhead.

Reply:
Also found this: https://github.com/dano/aioprocessing

Doesn't look super promising, but if nothing else we could lift limited implementations of the necessary components from there.

Reply:
> What I couldn't figure out was how to get the friendly API that multiprocessing.Connection has of automatic pickling and unpickling of objects 

Ah, yes. `multiprocessing` is better for now, then.

> https://github.com/dano/aioprocessing
>
> Doesn't look super promising, but if nothing else we could lift limited implementations of the necessary components from there.

Yeah, a failing build and no activity for a year is a no-go for direct usage. Since they call out `multiprocessing.Connection` specifically, it seems worth a look at some point. Still doesn't seem urgent, though.